libsolidity/analysis/TypeChecker.cpp

```diff
@@ -3768,14 +3769,22 @@ bool TypeChecker::visit(Literal const& _literal)
 	if (!subDenomination && literalRationalType)
 	{
 		auto&& [mantissa, exponent] = literalRationalType->mantissaExponent();
-		solAssert((mantissa && exponent) || (!mantissa && !exponent));
-		if (!mantissa)
-			m_errorReporter.typeError(
-				5503_error,
-				_literal.location(),
-				"This fractional number cannot be decomposed into a mantissa and decimal exponent "
-				"that fit the range of parameters of any possible suffix function."
-			);
+
+		if (!mantissa || !exponent)
+		{
+			string mantissaOrExponentErrorMessage = "This fractional number cannot be decomposed into a ";
+
+			if (!mantissa && ! exponent)
+				mantissaOrExponentErrorMessage += "mantissa and exponent that fit ";
+			else if (!exponent)
+				mantissaOrExponentErrorMessage += "exponent that fits ";
+			else if (!mantissa)
+				mantissaOrExponentErrorMessage += "mantissa that fits ";
+
+			mantissaOrExponentErrorMessage += "the range of parameters of any possible suffix function.";
```

Comment:
This is a bit of a nit, but are we allowed to use `fmt::format` here, as it would make this a little bit less verbose?

Reply:
I guess we can use it since I've found uses of it on other parts of the code. I was not pleased by this `if ... else` form too. Thanks!

Reply:
You're still gonna need the `if else` though :D

Reply:
Ah, right!
I guess you are suggesting something like this:
```suggestion
	string mantissaOrExponentErrorMessage;

	if (!mantissa && ! exponent)
		mantissaOrExponentErrorMessage = "mantissa and exponent that fit";
	else if (!exponent)
		mantissaOrExponentErrorMessage = "exponent that fits";
	else if (!mantissa)
		mantissaOrExponentErrorMessage = "mantissa that fits";

	m_errorReporter.typeError(
		5503_error,
		_literal.location(),
		fmt::format(
			"This fractional number cannot be decomposed into a {} "
			"the range of parameters of any possible suffix function.",
			mantissaOrExponentErrorMessage
		)
	);
```

Reply:
Exactly, plus the last `else if (!mantissa)` can be just an else, since by the time you get to the branch, you're left with only one path.

Reply:
Yeah, good point!
Thanks! This feels cleaner now.